### PR TITLE
BinaryFormatter disabled & obsoletion of legacy serialization APIs

### LINKED ISSUE #34891: [Breaking change]: BinaryFormatter disabled across almost all project types by default
### Description

Beginning with **.NET 8 Preview 4**, the `BinaryFormatter.Serialize` and `BinaryFormatter.Deserialize` methods will throw `NotSupportedException` across nearly all project types, including console applications.

A compatibility switch exists for applications which need to reenable BinaryFormatter behavior. See below for details.

### Version

.NET 8 Preview 4

### Previous behavior

In .NET 7, the `BinaryFormatter.Serialize` and `BinaryFormatter.Deserialize` methods were marked obsolete as error. However, if an application suppressed the obsoletion, the methods could still be called and would still function properly in most project types (excluding aspnet, wasm, and MAUI). For example, the APIs would function properly within a console app.

### New behavior

In .NET 8 Preview 4, these methods will throw `NotSupportedException` at runtime across all project types except WinForms and WPF. The APIs continue to remain obsolete-as-error across _all_ project types, including WinForms and WPF.

### Type of breaking change

- [ ] **Binary incompatible**: Existing binaries may encounter a breaking change in behavior, such as failure to load or execute, and if so, require recompilation.
- [ ] **Source incompatible**: When recompiled using the new SDK or component or to target the new runtime, existing source code may require source changes to compile successfully.
- [X] **Behavioral change**: Existing binaries may behave differently at run time.

### Reason for change

This is the next stage of the [BinaryFormatter obsoletion plan](https://github.com/dotnet/designs/blob/main/accepted/2020/better-obsoletion/binaryformatter-obsoletion.md), preparing for BinaryFormatter's eventual removal from .NET. See earlier breaking change notifications for additional context:

- [.NET 5 BinaryFormatter breaking changes](https://learn.microsoft.com/dotnet/core/compatibility/core-libraries/5.0/binaryformatter-serialization-obsolete)
- [.NET 7 BinaryFormatter breaking changes](https://learn.microsoft.com/otnet/core/compatibility/core-libraries/7.0/binaryformatter-apis-produce-errors)

### Recommended action

**The best course of action is to migrate away from BinaryFormatter due to its security and reliability flaws.**

However, should you need to continue using BinaryFormatter, please review [the *Recommended action* section](https://learn.microsoft.com/dotnet/core/compatibility/core-libraries/7.0/binaryformatter-apis-produce-errors#recommended-action) of the .NET 7 breaking change notification. That document provides a compatibility switch that can be set in your application's project file to reenable BinaryFormatter functionality. That compatibility switch continues to be honored in .NET 8.

### Feature area

Serialization

### Affected APIs

* [`System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize(Stream, object)`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.formatters.binary.binaryformatter.serialize)
* [`System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize(Stream)`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.formatters.binary.binaryformatter.deserialize)

### LINKED ISSUE #34893: [Breaking change]: Legacy serialization infrastructure APIs marked obsolete
### Description

Beginning with **.NET 8 Preview 4**, most of the legacy serialization infrastructure (the infrastructure which supports `SerializableAttribute` and `ISerializable`, along with infrastructure which supports `BinaryFormatter`) is obsolete.

Additionally, the entirety of the type `BinaryFormatter` is now obsolete _as error_.

(See related breaking change notification https://github.com/dotnet/docs/issues/34891.)

### Version

.NET 8 Preview 4

### Previous behavior

In .NET 7, the `Serialize` and `Deserialize` methods on `BinaryFormatter`, `IFormatter`, and `Formatter` were marked obsolete. The types themselves, however, were not marked obsolete. The full list of APIs marked obsolete in .NET 7 is provided at [the .NET 7 breaking change notification document](https://learn.microsoft.com/dotnet/core/compatibility/core-libraries/7.0/binaryformatter-apis-produce-errors).

### New behavior

Beginning with .NET 8 Preview 4, the entirety of the `BinaryFormatter`, `IFormatter`, and `Formatter` types are marked obsolete _as error_. Code which references these types - even if it does not call the `Serialize` or `Deserialize` method - will observe compilation failures on .NET 8.

Additionally, beginning with .NET 8 Preview 4, many types and APIs which support the legacy serialization infrastructure are marked obsolete _as warning_. Other APIs are marked `[EditorBrowsable(EditorBrowsableState.Never)]`, which causes these APIs to be hidden from the Visual Studio IDE.

See below for resolution strategies, including getting existing code which relied on `BinaryFormatter` to compile successfully.

### Type of breaking change

- [ ] **Binary incompatible**: Existing binaries may encounter a breaking change in behavior, such as failure to load or execute, and if so, require recompilation.
- [X] **Source incompatible**: When recompiled using the new SDK or component or to target the new runtime, existing source code may require source changes to compile successfully.
- [ ] **Behavioral change**: Existing binaries may behave differently at run time.

### Reason for change

This is the next stage of the [BinaryFormatter obsoletion plan](https://github.com/dotnet/designs/blob/main/accepted/2020/better-obsoletion/binaryformatter-obsoletion.md), preparing for BinaryFormatter's eventual removal from .NET. See earlier breaking change notifications for additional context:

- [.NET 5 BinaryFormatter breaking changes](https://learn.microsoft.com/dotnet/core/compatibility/core-libraries/5.0/binaryformatter-serialization-obsolete)
- [.NET 7 BinaryFormatter breaking changes](https://learn.microsoft.com/dotnet/core/compatibility/core-libraries/7.0/binaryformatter-apis-produce-errors)

### If you're using _BinaryFormatter_

The best course of action is to migrate away from it due to its security and reliability flaws. See https://aka.ms/binaryformatter for more information. If necessary, you can suppress the compilation error by following the steps in [the _Recommended action_ section](https://learn.microsoft.com/dotnet/core/compatibility/core-libraries/7.0/binaryformatter-apis-produce-errors#recommended-action) of the .NET 7 breaking change notification.

### If you're using _FormatterServices.GetUninitializedObject_

Use [`RuntimeHelpers.GetUninitializedObject`](https://learn.microsoft.com/dotnet/api/system.runtime.compilerservices.runtimehelpers.getuninitializedobject) instead.

If you're cross-compiling for .NET Framework and modern .NET, you can use an `#if` statement to selectively call the appropriate API, as shown below.

```cs
Type typeToInstantiate;
#if NET5_0_OR_GREATER
object obj = System.Runtime.CompilerServices.RuntimeHelpers.GetUninitializedObject(typeToInstantiate);
#else
object obj = System.Runtime.Serialization.FormatterServices.GetUninitializedObject(typeToInstantiate);
#endif
```

### If you're creating a custom _System.Exception_-derived type

Consider whether you truly need your custom exception type to be serializable. Chances are you do not need it to be serializable, as exception serialization is primarily intended to support remoting, and support for remoting was dropped in .NET Core 1.0.

If you have defined your custom exception type like this:

```cs
[Serializable]
public class MyException : Exception
{
    public MyException() { /* your ctor logic here */ }
    public MyException(string message) : base(message) { /* your ctor logic here */ }
    public MyException(string message, Exception inner) : base(message, inner) { /* your ctor logic here */ }
    protected MyException(SerializationInfo info, StreamingContext context)
        : base(info, context)
    {
        /* your rehydration logic here */
    }

    public override void GetObjectData(SerializationInfo info, StreamingContext context)
    {
        /* your dehydration here */

        base.GetObjectData(info, context);
    }
}
```

Consider simply removing the `[Serializable]` attribute, the serialization constructor, and the `GetObjectData` method override, as shown below.

```cs
// [Serializable] <-- Remove this attribute.
public class MyException : Exception
{
    public MyException() { /* your ctor logic here */ }
    public MyException(string message) : base(message) { /* your ctor logic here */ }
    public MyException(string message, Exception inner) : base(message, inner) { /* your ctor logic here */ }

    // Remove the constructor below.
    // protected MyException(SerializationInfo info, StreamingContext context)
    //     : base(info, context)
    // {
    //     /* your rehydration logic here */
    // }

    // Remove the method below.
    // public override void GetObjectData(SerializationInfo info, StreamingContext context)
    // {
    //     /* your dehydration here */
    // 
    //     base.GetObjectData(info, context);
    // }
}
```

There may be some cases where you cannot remove these APIs from your custom exception types. This might occur if you produce a library constrained by API compatibility requirements. In this case, the recommendation is to obsolete your own serialization constructor and `GetObjectData` methods using the `SYSLIB0051` diagnostic code, as shown below. Since ideally nobody outside the serialization infrastructure itself should be calling these APIs, this should only impact other types which subclass your custom exception type. It should not virally impact anybody catching, constructing, or otherwise using your custom exception type.

```cs
[Serializable]
public class MyException : Exception
{
    public MyException() { /* your ctor logic here */ }
    public MyException(string message) : base(message) { /* your ctor logic here */ }
    public MyException(string message, Exception inner) : base(message, inner) { /* your ctor logic here */ }

    [Obsolete(DiagnosticId = "SYSLIB0051")] // add this attribute to the serialization ctor
    protected MyException(SerializationInfo info, StreamingContext context)
        : base(info, context)
    {
        /* your rehydration logic here */
    }

    [Obsolete(DiagnosticId = "SYSLIB0051")] // add this attribute to GetObjectData
    public override void GetObjectData(SerializationInfo info, StreamingContext context)
    {
        /* your dehydration here */

        base.GetObjectData(info, context);
    }
}
```

If you're cross-targeting for .NET Framework and .NET 8+, you can use an `#if` statement to apply the obsoletion conditionally. This is the same strategy we use within the .NET libraries code base when we're cross-targeting runtimes.

```cs
[Serializable]
public class MyException : Exception
{
    public MyException() { /* your ctor logic here */ }
    public MyException(string message) : base(message) { /* your ctor logic here */ }
    public MyException(string message, Exception inner) : base(message, inner) { /* your ctor logic here */ }

#if NET8_0_OR_GREATER
    [Obsolete(DiagnosticId = "SYSLIB0051")] // add this attribute to the serialization ctor
#endif
    protected MyException(SerializationInfo info, StreamingContext context)
        : base(info, context)
    {
        /* your rehydration logic here */
    }

#if NET8_0_OR_GREATER
    [Obsolete(DiagnosticId = "SYSLIB0051")] // add this attribute to GetObjectData
#endif
    public override void GetObjectData(SerializationInfo info, StreamingContext context)
    {
        /* your dehydration here */

        base.GetObjectData(info, context);
    }
}
```

### If you're writing your own standalone serializable type

If you're declaring a custom type which is marked `[Serializable]` or which implements the `ISerializable` interface, you should not see any compilation warnings as a result of this change unless you've implemented a less common code pattern. For example, the code below should not see any warnings.

```cs
[Serializable]
public class Person
{
    private string name;
    private DateTime dateOfBirth;

    public string Name { get { return name; } set { name = value; } }
    public DateTime DateOfBirth { get { return dateOfBirth; } set { dateOfBirth = value; } }
}
```

However, .NET recommends _against_ annotating your types in this manner, as modern serialization libraries do not require such annotations. Instead, consider leaving off the `[Serializable]` attribute and the `ISerializable` interface, instead relying on your serialization library to access your object through its public properties rather than its private fields.

```cs
// [Serializable] <-- suggest removing this attribute
public class Person
{
    private string name;
    private DateTime dateOfBirth;

    public string Name { get { return name; } set { name = value; } }
    public DateTime DateOfBirth { get { return dateOfBirth; } set { dateOfBirth = value; } }
}
```

If you're subclassing an existing .NET type which is marked `[Serializable]` and you're observing `SYSLIB0051` warning codes, see the earlier section on creating custom exception types. The overall pattern described there remains generally applicable.

### If you're writing a serialization library

Microsoft strongly recommends _against_ serialization libraries supporting the legacy serialization infrastructure (`[Serializable]` and `ISerializable`). Modern serialization libraries should have policy based on a type's public APIs rather than its private implementation details. Basing a serializer on these implementation details and strongly tying it to `ISerializable` and other mechanisms which encourage embedding type names within the serialized payload can lead to many of the same problems called out in https://aka.ms/binaryformatter.

If your serialization library must remain compatible with the legacy serialization infrastructure, you can easily suppress the legacy serialization API obsoletions (the `SYSLIB0050` category) project-wide by putting the following block in your `.csproj`, `.vbproj`, or directory-level `.props` file.

```xml
<PropertyGroup>
  <NoWarn>$(NoWarn);SYSLIB0050</NoWarn>
</PropertyGroup>
```

### Feature area

Serialization

### Affected APIs

The following types are marked obsolete _as warning_ with obsoletion code `SYSLIB0050`.

- [`System.Runtime.Serialization.FormatterConverter`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.formatterconverter)
- [`System.Runtime.Serialization.FormatterServices`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.formatterservices)
- [`System.Runtime.Serialization.IFormatterConverter`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.iformatterconverter)
- [`System.Runtime.Serialization.IObjectReference`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.iobjectreference)
- [`System.Runtime.Serialization.ISafeSerializationData`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.isafeserializationdata)
- [`System.Runtime.Serialization.ISerializationSurrogate`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.iserializationsurrogate)
- [`System.Runtime.Serialization.ISurrogateSelector`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.isurrogateselector)
- [`System.Runtime.Serialization.ObjectIDGenerator`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.objectidgenerator)
- [`System.Runtime.Serialization.ObjectManager`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.objectmanager)
- [`System.Runtime.Serialization.SafeSerializationEventArgs`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.safeserializationeventargs)
- [`System.Runtime.Serialization.SerializationObjectManager`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.serializationobjectmanager)
- [`System.Runtime.Serialization.StreamingContextStates`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.streamingcontextstates)
- [`System.Runtime.Serialization.SurrogateSelector`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.surrogateselector)
- [`System.Runtime.Serialization.Formatters.FormatterAssemblyStyle`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.formatters.formatterassemblystyle)
- [`System.Runtime.Serialization.Formatters.FormatterTypeStyle`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.formatters.formattertypestyle)
- [`System.Runtime.Serialization.Formatters.IFieldInfo`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.formatters.ifieldinfo)
- [`System.Runtime.Serialization.Formatters.TypeFilterLevel`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.formatters.typefilterlevel)

The following APIs are marked obsolete _as warning_ with obsoletion code `SYSLIB0050`.

- [`System.Type.IsSerializable`](https://learn.microsoft.com/dotnet/api/system.type.isserializable) property (and all overridden implementations)
- [`System.Reflection.FieldAttributes.NotSerialized`](https://learn.microsoft.com/dotnet/api/system.reflection.fieldattributes) enum value (but not the `FieldAttributes` enum type itself)
- [`System.Reflection.FieldInfo.IsNotSerialized`](https://learn.microsoft.com/dotnet/api/system.reflection.fieldinfo.isnotserialized) property
- [`System.Reflection.TypeAttributes.Serializable`](https://learn.microsoft.com/dotnet/api/system.reflection.typeattributes) enum value (but not the `TypeAttributes` enum type itself)
- [`System.Runtime.Serialization.ISerializable.GetObjectData`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.iserializable.getobjectdata) method (but not the `ISerializable` type itself)
- All constructors of [`System.Runtime.Serialization.SerializationInfo`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.serializationinfo.-ctor) (but not the `SerializationInfo` type itself)
- All constructors of [`System.Runtime.Serialization.StreamingContext`](https://learn.microsoft.com/dotnet/api/system.runtime.serialization.streamingcontext.-ctor) (but not the `StreamingContext` type itself)

The following APIs are marked obsolete _as warning_ with obsoletion code `SYSLIB0051`.

* All public or protected serialization constructors. These are constructors which follow the pattern `.ctor(SerializationInfo, StreamingContext)`. An example of such a constructor is [the serialization constructor on the `System.Exception` type](https://learn.microsoft.com/dotnet/api/system.exception.-ctor#system-exception-ctor(system-runtime-serialization-serializationinfo-system-runtime-serialization-streamingcontext)).
* All implicit implementations of the `IObjectReference.GetRealObject` method. An example of such a method is [`System.Reflection.ParameterInfo.GetRealObject`](https://learn.microsoft.com/dotnet/api/system.reflection.parameterinfo.getrealobject).
* All implicit implementations of the `ISerializable.GetObjectData` method. An example of such a method is [`System.Exception.GetObjectData`](https://learn.microsoft.com/dotnet/api/system.exception.getobjectdata).

The following types are marked `[EditorBrowsable(EditorBrowsableState.Never)]`, but they _are not_ obsoleted. Marking them with the `[EditorBrowsable(EditorBrowsableState.Never)]` annotation will hide them from within the Visual Studio IDE. However, they can still be referenced by .NET 8 apps, and callers will not observe a compilation warning or error when referencing these types.

- [`System.NonSerializedAttribute`](https://learn.microsoft.com/dotnet/api/system.nonserializedattribute)
- [`System.SerializableAttribute`](https://learn.microsoft.com/dotnet/api/system.serializableattribute)

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #34891
Fixes #34893

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/compatibility/core-libraries/8.0/obsolete-apis-with-custom-diagnostics.md](https://github.com/dotnet/docs/blob/90f932bef24ddb5c90ab2986e408600a8bde7a0d/docs/core/compatibility/core-libraries/8.0/obsolete-apis-with-custom-diagnostics.md) | [Obsolete APIs with custom diagnostics - .NET 8](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/8.0/obsolete-apis-with-custom-diagnostics?branch=pr-en-us-35217) |
| [docs/core/compatibility/serialization/8.0/binaryformatter-disabled.md](https://github.com/dotnet/docs/blob/90f932bef24ddb5c90ab2986e408600a8bde7a0d/docs/core/compatibility/serialization/8.0/binaryformatter-disabled.md) | [BinaryFormatter is disabled](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/serialization/8.0/binaryformatter-disabled?branch=pr-en-us-35217) |
| [docs/fundamentals/syslib-diagnostics/syslib0011.md](https://github.com/dotnet/docs/blob/90f932bef24ddb5c90ab2986e408600a8bde7a0d/docs/fundamentals/syslib-diagnostics/syslib0011.md) | [SYSLIB0011: BinaryFormatter serialization is obsolete](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/syslib0011?branch=pr-en-us-35217) |
| [docs/fundamentals/syslib-diagnostics/syslib0050.md](https://github.com/dotnet/docs/blob/90f932bef24ddb5c90ab2986e408600a8bde7a0d/docs/fundamentals/syslib-diagnostics/syslib0050.md) | [SYSLIB0050](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/syslib0050?branch=pr-en-us-35217) |
| [docs/fundamentals/syslib-diagnostics/syslib0051.md](https://github.com/dotnet/docs/blob/90f932bef24ddb5c90ab2986e408600a8bde7a0d/docs/fundamentals/syslib-diagnostics/syslib0051.md) | [SYSLIB0051](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/syslib0051?branch=pr-en-us-35217) |

---

#### Internal previews

<details><summary><strong>Toggle expand/collapse</strong></summary><br/>

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/compatibility/5.0.md](https://github.com/dotnet/docs/blob/c9d542896181192f7e18da22945a9e725bbc2305/docs/core/compatibility/5.0.md) | [Breaking changes in .NET 5](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/5.0?branch=pr-en-us-35217) |
| [docs/core/compatibility/7.0.md](https://github.com/dotnet/docs/blob/c9d542896181192f7e18da22945a9e725bbc2305/docs/core/compatibility/7.0.md) | [Breaking changes in .NET 7](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/7.0?branch=pr-en-us-35217) |
| [docs/core/compatibility/8.0.md](https://github.com/dotnet/docs/blob/c9d542896181192f7e18da22945a9e725bbc2305/docs/core/compatibility/8.0.md) | [docs/core/compatibility/8.0](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/8.0?branch=pr-en-us-35217) |
| [docs/core/compatibility/core-libraries/8.0/obsolete-apis-with-custom-diagnostics.md](https://github.com/dotnet/docs/blob/c9d542896181192f7e18da22945a9e725bbc2305/docs/core/compatibility/core-libraries/8.0/obsolete-apis-with-custom-diagnostics.md) | [docs/core/compatibility/core-libraries/8.0/obsolete-apis-with-custom-diagnostics](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/8.0/obsolete-apis-with-custom-diagnostics?branch=pr-en-us-35217) |
| [docs/core/compatibility/serialization/5.0/binaryformatter-serialization-obsolete.md](https://github.com/dotnet/docs/blob/c9d542896181192f7e18da22945a9e725bbc2305/docs/core/compatibility/serialization/5.0/binaryformatter-serialization-obsolete.md) | [docs/core/compatibility/serialization/5.0/binaryformatter-serialization-obsolete](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/serialization/5.0/binaryformatter-serialization-obsolete?branch=pr-en-us-35217) |
| [docs/core/compatibility/serialization/7.0/binaryformatter-apis-produce-errors.md](https://github.com/dotnet/docs/blob/c9d542896181192f7e18da22945a9e725bbc2305/docs/core/compatibility/serialization/7.0/binaryformatter-apis-produce-errors.md) | [docs/core/compatibility/serialization/7.0/binaryformatter-apis-produce-errors](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/serialization/7.0/binaryformatter-apis-produce-errors?branch=pr-en-us-35217) |
| [docs/core/compatibility/serialization/7.0/serializationformat-binary.md](https://github.com/dotnet/docs/blob/c9d542896181192f7e18da22945a9e725bbc2305/docs/core/compatibility/serialization/7.0/serializationformat-binary.md) | [docs/core/compatibility/serialization/7.0/serializationformat-binary](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/serialization/7.0/serializationformat-binary?branch=pr-en-us-35217) |
| [docs/core/compatibility/serialization/8.0/binaryformatter-disabled.md](https://github.com/dotnet/docs/blob/c9d542896181192f7e18da22945a9e725bbc2305/docs/core/compatibility/serialization/8.0/binaryformatter-disabled.md) | [docs/core/compatibility/serialization/8.0/binaryformatter-disabled](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/serialization/8.0/binaryformatter-disabled?branch=pr-en-us-35217) |
| [docs/core/compatibility/unsupported-apis.md](https://github.com/dotnet/docs/blob/c9d542896181192f7e18da22945a9e725bbc2305/docs/core/compatibility/unsupported-apis.md) | [APIs that always throw exceptions on .NET Core and .NET 5+](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/unsupported-apis?branch=pr-en-us-35217) |
| [docs/core/deploying/trimming/trimming-options.md](https://github.com/dotnet/docs/blob/c9d542896181192f7e18da22945a9e725bbc2305/docs/core/deploying/trimming/trimming-options.md) | [Trimming options](https://review.learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options?branch=pr-en-us-35217) |
| [docs/fundamentals/syslib-diagnostics/obsoletions-overview.md](https://github.com/dotnet/docs/blob/c9d542896181192f7e18da22945a9e725bbc2305/docs/fundamentals/syslib-diagnostics/obsoletions-overview.md) | [docs/fundamentals/syslib-diagnostics/obsoletions-overview](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/obsoletions-overview?branch=pr-en-us-35217) |
| [docs/fundamentals/syslib-diagnostics/syslib0011.md](https://github.com/dotnet/docs/blob/c9d542896181192f7e18da22945a9e725bbc2305/docs/fundamentals/syslib-diagnostics/syslib0011.md) | [SYSLIB0011: BinaryFormatter serialization is obsolete](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/syslib0011?branch=pr-en-us-35217) |
| [docs/fundamentals/syslib-diagnostics/syslib0038.md](https://github.com/dotnet/docs/blob/c9d542896181192f7e18da22945a9e725bbc2305/docs/fundamentals/syslib-diagnostics/syslib0038.md) | [SYSLIB0038: SerializationFormat.Binary is obsolete](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/syslib0038?branch=pr-en-us-35217) |
| [docs/fundamentals/syslib-diagnostics/syslib0050.md](https://github.com/dotnet/docs/blob/c9d542896181192f7e18da22945a9e725bbc2305/docs/fundamentals/syslib-diagnostics/syslib0050.md) | [docs/fundamentals/syslib-diagnostics/syslib0050](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/syslib0050?branch=pr-en-us-35217) |
| [docs/fundamentals/syslib-diagnostics/syslib0051.md](https://github.com/dotnet/docs/blob/c9d542896181192f7e18da22945a9e725bbc2305/docs/fundamentals/syslib-diagnostics/syslib0051.md) | [docs/fundamentals/syslib-diagnostics/syslib0051](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/syslib0051?branch=pr-en-us-35217) |

</details>

### CODE REVIEW

File: docs/core/compatibility/serialization/8.0/binaryformatter-disabled.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+## Recommended action
+
+**The best course of action is to migrate away from `BinaryFormatter` due to its security and reliability flaws.**
+
+However, should you need to continue using `BinaryFormatter`, you can set a compatibility switch in your project file to re-enable `BinaryFormatter` functionality. For more information, see the [Recommended action](../7.0/binaryformatter-apis-produce-errors.md#recommended-action) section of the .NET 7 breaking change notification. That compatibility switch continues to be honored in .NET 8.
+
+## Affected APIs
+
+- <xref:System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize(System.IO.Stream,System.Object)?displayProperty=fullName>
```
Comment (gewarren) [inline docs/core/compatibility/serialization/8.0/binaryformatter-disabled.md:38]: @GrabYourPitchforks Is it just these two specific overloads, or all overloads?

Comment (GrabYourPitchforks) [reply docs/core/compatibility/serialization/8.0/binaryformatter-disabled.md:38]: Just those two methods. Other overloads (like `public object Deserialize(Stream serializationStream, HeaderHandler handler)`) don't exist at all in .NET Core, so they're not candidates for obsoletion.

**Review by BillWagner — APPROVED** [review]
This all LGTM @gewarren 

You can :shipit: when ready.

**Comment by gewarren** [comment]
Thanks so much @BillWagner

**Pull request merged**